File: LinkedList/linked_list.py
class Node():
    def __init__(self,data) :
        self.data = data
        self.next  = None

    # def __repr__(self) -> str:
    #     return f"Data:{self.data}, Next:({self.next})"




# added a node
def add(data,head):
    new_node = Node(data=data)
    if head==None:
        head = new_node
        head.next = None
    else:   
        new_node.next=head
        head = new_node
    return head

#remove a node 
def rem(data,head):
    current = head
    while(data!=current.data):
        prev = current
        current = current.next
        
        if (data==current.data):
            prev.next=current.next
            break
        
        
        if (head.next==None):
            print("Unable to find the element")
            break
    
    
#Insert a element
def ins(data,pos,head):
    new_node = Node(data)
    
    current = head
    i =1
    while (current.next!=None and pos!=0):
       
        if (i==pos-1):
            temp = current.next
            current.next = new_node
            new_node.next = temp
            break
        else:
            current = current.next

            i+=1
        


def printList(head):
   
    current = head
    
    while (current!=None):
        print("current: " ,current , "current.next: ",current.next, "current.data: ",current.data, end="\n")
        
        current = current.next
   
              


    

def ui():
    print("\nWhat function you want to perform.")
    print("\t1. ADD to linkedlist")
    print("\t2. REMOVE from LinkedList")
    print("\t3. INSERT LinkedList")
    print("\t4. PRINT LinkedList")
    print("\t5. QUIT")
    

option = -1
head =None
while (option !=5):
    ui()
    option = int(input(""))
    if (option>=1 and option<=5):
        if option == 1:
            print("Data to add into the linked list")
            data = int(input(""))
            head = add(data,head)

        elif option == 2:
            
            print("Which no. to remove")
            data = int(input(""))
            rem(data,head)

            

        elif option == 3:
            print("Which no. to insert")
            data = int(input(""))
            pos= int(input("Position:"))
            ins(data,pos,head)

        elif option == 4:
            
            printList(head)
        
               

File: LinkedList/linked_list.c
#include <stdio.h>
#include <stdlib.h>

typedef struct  
{
    int data;
    void *next;

} Node;

Node *head = NULL;


// add to list

Node *addNode(int data){
    Node *new = malloc(sizeof(Node));
    if (head == NULL){
        if (new==NULL) return NULL;
        new->data = data;
        head = new;
        new->next = NULL;
    }else{
        if (new==NULL) return NULL;
        new->data = data;
        new->next = head;
        head = new;


    }
}
//remove a node from list
int removeNode(int data){
    Node *current = head;
    Node *prev = head;
    while (current!=NULL){
        if (current->data == data){
             // if current node is head 
            if (current == head ) head = current->next;
            else {
                prev ->next = current->next; 
            }
            return 1;
        }
        prev = current;
        current= current->next;
    }
    return 0 ;

}
// insert a node in a list
void insertNode(int data, int place){
    Node *insert = malloc(sizeof(Node));
    Node *current = head;
    insert->data = data;
    int count = 1;
    while (current!=NULL && place !=0){
        if (count == place-1){
            int * temp =current->next;
            current->next = insert;
            insert->next =  temp;
            break;
        }else 
        {
        current= current->next;}
        count ++;

    }
    if (place!=0){
        printf("not a good position ");
    }
}
// print the list 
void printlist(){
    Node *current = head ;
    while (current!=NULL){
        printf(" %d -> ",current->data);
        current = current->next;
    }
    printf("\n");
}

// ui
void printMenu(){
    printf("Following things which u can do\n");
    printf("\t 1. ADD something\n");
    printf("\t 2. REMOVE something\n");
    printf("\t 3. INSERT something\n");
    printf("\t 4. PRINT list\n");
    printf("\t 5. QUIT\n");
}

int main(){


    int option =-1;
    while (option!=5){
        printMenu();
        int num = scanf("%d",&option);
        if (num ==1 && option>=1 && option<=5){
            switch (option)
            {
            case 1:
                printf("What to add?\n");
                scanf("%d",&option);
                
                Node *new = addNode(option);
                break;
            case 2:
                // rem
                printf("Which no. to remove\n");
                scanf("%d",&option);
                int succes= removeNode(option);
                if (!succes) printf( "element not found");

                break;

            case 3:
                // insert
                printf (" what to insert and where to insert\n");
                scanf("%d",&option);
                int place ;
                scanf("%d",&place);
                insertNode(option,place);

                break;
            
            case 4:
                printlist();
            case 5:
                break;
            default:
                printf("Invalid option\n");
            }
        }
    }
    return 0;
}